Prefect Fixes

pgierz opened the pull request:
This pull request introduces several changes to enhance the functionality and clarity of the `pymorize` project. The key updates include adding new commands to the CLI, refactoring methods for better readability, and improving the handling of Prefect workflows.

### CLI Enhancements:
* Added a new command `prefect_check` to the CLI for checking Prefect with a dummy flow using a specified configuration file (`src/pymorize/cli.py`).

### Method Refactoring:
* Renamed several methods in `CMORizer` to improve clarity and consistency, such as `_post_init_data_request_variables` to `_post_init_populate_rules_with_data_request_variables` and `_post_init_read_dimensionless_unit_mappings` to `_post_init_populate_rules_with_dimensionless_unit_mappings` (`src/pymorize/cmorizer.py`). [[1]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aL124-R125) [[2]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aL236-R236) [[3]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aL249-R249) [[4]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aL485-R490)

### Prefect Workflow Improvements:
* Introduced the `_match_pipelines_in_rules` method to the `CMORizer` class to match pipelines within rules and updated the `process` method to call this new method (`src/pymorize/cmorizer.py`). [[1]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aR281-R284) [[2]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aR559)
* Added Prefect flows and tasks for caching checks and rule processing in the `CMORizer` class, including the `check_prefect` and `_caching_check` methods (`src/pymorize/cmorizer.py`). [[1]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aL618-L633) [[2]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aL583-R592)

### Pipeline Enhancements:
* Implemented custom pickling for the `Pipeline` class to handle Prefect steps and cluster assignments, along with a new `_steps_are_prefectized` attribute (`src/pymorize/pipeline.py`). [[1]](diffhunk://#diff-899aef304e213d8841d53cf1be180288fd2fec7a9421370d233b5d437c43bb54R34) [[2]](diffhunk://#diff-899aef304e213d8841d53cf1be180288fd2fec7a9421370d233b5d437c43bb54R61-R86) [[3]](diffhunk://#diff-899aef304e213d8841d53cf1be180288fd2fec7a9421370d233b5d437c43bb54R101-R102)

### Code Cleanup:
* Removed deprecated and unused code in the `Rule` class, including old properties and commented-out methods (`src/pymorize/rule.py`). [[1]](diffhunk://#diff-2c62d44942e58feeeb6acdb6b358409aabb2901bf54c747b053a42fcebc3d778L69-L78) [[2]](diffhunk://#diff-2c62d44942e58feeeb6acdb6b358409aabb2901bf54c747b053a42fcebc3d778L138-L144) [[3]](diffhunk://#diff-2c62d44942e58feeeb6acdb6b358409aabb2901bf54c747b053a42fcebc3d778L216-L225) [[4]](diffhunk://#diff-2c62d44942e58feeeb6acdb6b358409aabb2901bf54c747b053a42fcebc3d778L284-L313)

pgierz commented on src/pymorize/pipeline.py:
```diff
@@ -56,12 +58,32 @@ def __str__(self):
             r_val.append(f"[{i+1}/{len(self.steps)}] {step.__name__}")
         return "\n".join(r_val)
 
+    def __getstate__(self):
+        """Custom pickling of a Pipeline"""
+        state = self.__dict__.copy()
+        if self._steps_are_prefectized:
+            state["_steps"] = self._raw_steps
+            del state["_raw_steps"]
+            state["_steps_are_prefectized"] = False
+        # It makes no sense to pickle the cluster
+        del state["_cluster"]
+
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        if self._workflow_backend == "prefect":
+            self._prefectize_steps()
+        logger.info("Restoring from pickled state!")
+        logger.info("You may want to assign a cluster to this pipeline")
+
```
This part was specific to cache `Rule` objects, since `Rule`s have the `Pipeline` attached to them, and the `Pipeline` in turn has a collection of Prefect `Task` objects. Also, it is not sensible to cache the Dask `SLURMCluster` object, since that is unique to each run.

I'm not sure which one of the two was causing it to be impossible to pickle things.

mandresm replied:
Thanks for pointing out the place. Looks like a good solution to me, since the pipeline tasks will be run anyway with their corresponding caches.

mandresm commented on src/pymorize/pipeline.py:
```diff
@@ -56,12 +58,32 @@ def __str__(self):
+        """Custom pickling of a Pipeline"""
+        state = self.__dict__.copy()
+        if self._steps_are_prefectized:
+            state["_steps"] = self._raw_steps
+            del state["_raw_steps"]
```
I think I get it now: you have non-prefectized steps in `self._raw_steps` and prefectized steps in `self._steps` so for the state you are overwriting the prefectize ones with the raw ones right? :)

pgierz replied:
Yes, exactly.

mandresm reviewed (approved):
Nice job, feel free to merge. The generator errors are also gone for me while running `sample.yaml` in levante. Only a couple of questions

pgierz commented on src/pymorize/pipeline.py:
```diff
@@ -56,12 +58,32 @@ def __str__(self):
+        if self._workflow_backend == "prefect":
+            self._prefectize_steps()
+        logger.info("Restoring from pickled state!")
+        logger.info("You may want to assign a cluster to this pipeline")
```
I guess for myself...Eventually there should be some logic here to attach the cluster during de-pickling, but I have no idea where it will be called so I'm not sure what to hook up here.

pgierz commented:
There's still one failing test which I need to figure out. Good that it is working "in real life", though. @mandresm, you didn't have any more problems, right? What version are you currently using?

mandresm commented:
> There's still one failing test which I need to figure out. Good that it is working "in real life", though. @mandresm, you didn't have any more problems, right? What version are you currently using?

My problems are related to disk space seemingly, but I did not have time to go into more details.

I was using this branch, commit id: `3f1b92bddf451dbc4d6b31d8dcdbbd1b25ef09cd`

pgierz commented:
...and it already bit me in the butt :-(

                    INFO     Restoring from pickled state!        pipeline.py:77
                    INFO     You may want to assign a cluster to  pipeline.py:78
                             this pipeline        

Later in the error logs:

Pipeline has no attribute `_cluster`.

Now I need to figure out how to interrupt the depickle process to inject parts into the newly recreated object.

pgierz commented:
This pull request introduces several enhancements and refactorings to the `pymorize` project, focusing on improving Dask cluster management, refining the CMORizer class, and updating pipeline and rule handling. The most important changes include the addition of a `DaskContext` class for managing Dask clusters, updates to the CMORizer class to better handle rules and workflows, and improvements to pipeline and rule serialization and deserialization.

### Cluster Management:
* Added `DaskContext` class to manage the current Dask cluster, ensuring only one active cluster at a time (`src/pymorize/cluster.py`).

### CMORizer Enhancements:
* Updated CMORizer class methods to use more descriptive names and handle rules more effectively (`src/pymorize/cmorizer.py`). [[1]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aL124-R126) [[2]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aL236-R237) [[3]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aR247-R252) [[4]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aR284-R287) [[5]](diffhunk://#diff-9deea47805fe528815a5e35850622c720823cb8c6abd547246359d9502aae47aL483-R491)
* Introduced a new `check_prefect` flow for validating workflows with Prefect (`src/pymorize/cmorizer.py`).

### Pipeline Improvements:
* Implemented custom pickling methods for the Pipeline class to handle Prefect steps and Dask clusters (`src/pymorize/pipeline.py`). [[1]](diffhunk://#diff-899aef304e213d8841d53cf1be180288fd2fec7a9421370d233b5d437c43bb54R62-R92) [[2]](diffhunk://#diff-899aef304e213d8841d53cf1be180288fd2fec7a9421370d233b5d437c43bb54R107-R108) [[3]](diffhunk://#diff-899aef304e213d8841d53cf1be180288fd2fec7a9421370d233b5d437c43bb54L98-R128)
* Added a deep copy operation in `_prefectize_steps` to ensure proper handling of Prefect tasks (`src/pymorize/pipeline.py`).

### Rule Handling:
* Simplified the Rule class by removing unnecessary properties and methods, and directly managing pipelines (`src/pymorize/rule.py`). [[1]](diffhunk://#diff-2c62d44942e58feeeb6acdb6b358409aabb2901bf54c747b053a42fcebc3d778L53-R50) [[2]](diffhunk://#diff-2c62d44942e58feeeb6acdb6b358409aabb2901bf54c747b053a42fcebc3d778L69-L78) [[3]](diffhunk://#diff-2c62d44942e58feeeb6acdb6b358409aabb2901bf54c747b053a42fcebc3d778L138-L144) [[4]](diffhunk://#diff-2c62d44942e58feeeb6acdb6b358409aabb2901bf54c747b053a42fcebc3d778L174-R154) [[5]](diffhunk://#diff-2c62d44942e58feeeb6acdb6b358409aabb2901bf54c747b053a42fcebc3d778L185-R165) [[6]](diffhunk://#diff-2c62d44942e58feeeb6acdb6b358409aabb2901bf54c747b053a42fcebc3d778L216-L225)

### CLI Updates:
* Added a new CLI command `prefect_check` to validate Prefect workflows using a given configuration file (`src/pymorize/cli.py`).
* Reorganized CLI subcommands for better structure and readability (`src/pymorize/cli.py`).
